src/ceph-volume/ceph_volume/devices/lvm/listing.py

```diff
@@ -70,51 +68,31 @@ def direct_report():
     bypasses the need to deal with the class interface which is meant for cli
     handling.
     """
-    _list = List([])
-    # this is crucial: make sure that all paths will reflect current
-    # information. In the case of a system that has migrated, the disks will
-    # have changed paths
-    _list.update()
-    return _list.full_report()
+    return List([]).full_report()
 
 
+# TODO: move somewhere in our codebase where it's more accessible/usable?
+def is_ceph_device(lv):
+    try:
+        lv.tags['ceph.osd_id']
+    except (KeyError, AttributeError):
+        logger.warning('device is not part of ceph: %s', lv)
+        return False
+
+    return True
+
+# TODO: Perhaps, get rid of this class and simplify this module further?
 class List(object):
 
     help = 'list logical volumes and devices associated with Ceph'
 
     def __init__(self, argv):
         self.argv = argv
 
-    @property
-    def pvs(self):
-        """
-        To avoid having to make an LVM API call for every single item being
-        reported, the call gets set only once, using that stored call for
-        subsequent calls
-        """
-        if getattr(self, '_pvs', None) is not None:
-            return self._pvs
-        self._pvs = api.get_api_pvs()
-        return self._pvs
-
-    def match_devices(self, lv_uuid):
-        """
-        It is possible to have more than one PV reported *with the same name*,
-        to avoid incorrect or duplicate contents we correlated the lv uuid to
-        the one on the physical device.
-        """
-        devices = []
-        for device in self.pvs:
-            if device.get('lv_uuid') == lv_uuid:
-                devices.append(device['pv_name'])
-        return devices
-
     @decorators.needs_root
     def list(self, args):
-        # ensure everything is up to date before calling out
-        # to list lv's
-        lvs = self.update()
-        report = self.generate(args, lvs)
+        report = self.single_report(args.device) if args.device else \
+                 self.full_report()
```

Comment:
Why not call `return self.create_report(api.get_lvs(), full_report=True)` right here and avoid the indirection through `full_report()`?

Reply:
This minor indirection can improve readability greatly in tests -
```
diff --git a/src/ceph-volume/ceph_volume/tests/devices/lvm/test_listing.py b/src/ceph-volume/ceph_volume/tests/devices/lvm/test_listing.py
index 0ca6992203..06a443f265 100644
--- a/src/ceph-volume/ceph_volume/tests/devices/lvm/test_listing.py
+++ b/src/ceph-volume/ceph_volume/tests/devices/lvm/test_listing.py
@@ -183,8 +183,10 @@ class TestFullReport(object):
                             pvolumes)
         monkeypatch.setattr(lvm.listing.api, 'get_lvs', lambda **kwargs:
                             volumes)
+        monkeypatch.setattr(lvm.listing.List, 'list',
+                            lvm.listing.List([]).create_report)
 
-        result = lvm.listing.List([]).full_report()
+        result = lvm.listing.List([]).list(volumes)
         assert result['0'][1]['path'] == '/dev/sda1'
         assert result['0'][1]['tags'] == {'PARTUUID': 'x'}
         assert result['0'][1]['type'] == 'wal'
[fedora31@fedora31 ceph_volume]$ 
```

Reply:
Number of arguments `create_report` can be reduced. I've made this change in recent push.

Reply:
My issue with this is how things are called here. There is no clear distinction which function does what. `list()` is the entry point and calls either `single_report()` or `full_report`. `single_report` retrieves what it wants to report about (a drive) and passes that to `create_report()`. `full_report()` however does nothing really and now `create_report` is responsible for retrieving the report data.
I think it would be cleaner if `create_report` only does one job...aggregate report data in this case, instead of also retrieving that data in some cases.
This kind of